File: server/models/avatar.go
package models

import (
	"math"
	"math/rand"
	"strings"
)

type Avatar string

// AvatarSpec is the info needed to create a correctly formatted avatar.
type AvatarSpec struct {
	Size    uint8    `json:"size"`
	Palette []string `json:"palette"`
}

// TODO: Probably a much better way of doing this but this should work fine for
//       now
func GenerateRandomAvatar(spec AvatarSpec) Avatar {
	pixels := make([]bool, spec.Size*spec.Size)
	sizeHalf := int(math.Ceil(float64(spec.Size) / 2))
	data := ""

	for y := 0; y < int(spec.Size); y++ {
		for x := 0; x < sizeHalf; x++ {
			pos := y*int(spec.Size) + x
			mirroredPos := y*int(spec.Size) + (int(spec.Size) - 1 - x)

			active := rand.Int31()%2 == 0

			pixels[pos] = active
			pixels[mirroredPos] = active
		}
	}

	for _, active := range pixels {
		if active {
			data += "1"
		} else {
			data += "0"
		}
	}

	data += spec.Palette[rand.Intn(len(spec.Palette))]
	return Avatar(data)
}

func (a Avatar) Validate(spec AvatarSpec) error {
	sections := strings.Split(string(a), "#")
	if len(sections) != 2 {
		return ErrInvalidAvatar
	}
	cellsString := sections[0]

	if len(cellsString) != int(spec.Size)*int(spec.Size) {
		return ErrInvalidAvatar
	}
	for _, cell := range cellsString {
		if cell != '1' && cell != '0' {
			return ErrInvalidAvatar
		}
	}
	color := sections[1]
	color = "#" + color

	for _, paletteColor := range spec.Palette {
		if color == paletteColor {
			return nil
		}
	}

	return ErrInvalidAvatar
}

File: client/src/utils/avatar.js
export function random(size) {
  const cells = [];

  for (let y = 0; y < size; y += 1) {
    for (let x = 0; x < Math.ceil(size / 2); x += 1) {
      const pos = y * size + x;
      const mirroredPos = y * size + (size - 1 - x);
      const cell = Math.random() >= 0.5;
      cells[pos] = cell;
      if (mirroredPos !== pos) {
        cells[mirroredPos] = cell;
      }
    }
  }

  return cells;
}

export function serialize(cells, color) {
  let data = cells.map((cell) => (cell ? "1" : "0")).join("");
  data += color;

  return data;
}

export function deserialize(data) {
  const [cellsString, colorString] = data.split("#");
  const size = Math.ceil(Math.sqrt(cellsString.length));
  const cells = [...cellsString].map((cell) => cell === "1");
  const color = `#${colorString}`;

  return [size, cells, color];
}
